source/slang-record-replay/replay/json-consumer.cpp

```diff
@@ -1,174 +1,1253 @@
+    }
+
+    void CommonInterfaceWriter::specialize(ObjectID objectId, slang::SpecializationArg const* specializationArgs,
+            SlangInt specializationArgCount, ObjectID outSpecializedComponentTypeId, ObjectID outDiagnosticsId)
+    {
+        Slang::StringBuilder builder;
+        int indent = 0;
+
+        Slang::String functionName = m_className;
+        functionName = functionName + "::specialize";
+
+        {
+            ScopeWritterForKey scopeWritter(&builder, &indent, functionName);
+            {
+                WRITE_PAIR(builder, indent, "this", Slang::StringUtil::makeStringWithFormat("0x%X", objectId));
+                if (specializationArgCount)
+                {
+                    ScopeWritterForKey scopeWritterForArgs(&builder, &indent, "specializationArgs", false);
+                    for(int i = 0; i < specializationArgCount; i++)
+                    {
+                        bool isLastField = (i == specializationArgCount - 1);
+                        ScopeWritterForKey scopeWritterForArg(&builder, &indent, Slang::StringUtil::makeStringWithFormat("[%d]\n", i), isLastField);
+                        {
+                            WRITE_PAIR(builder, indent, "kind", SpecializationArgKindToString(specializationArgs[i].kind));
+                            WRITE_PAIR_NO_COMMA(builder, indent, "type", Slang::StringUtil::makeStringWithFormat("0x%X", specializationArgs[i].type));
+                        }
+                    }
+                }
+                else
+                {
+                    WRITE_PAIR(builder, indent, "specializationArgs", "nullptr");
+                }
+                WRITE_PAIR(builder, indent, "specializationArgCount", specializationArgCount);
+                WRITE_PAIR(builder, indent, "outSpecializedComponentType", Slang::StringUtil::makeStringWithFormat("0x%X", outSpecializedComponentTypeId));
+                WRITE_PAIR_NO_COMMA(builder, indent, "outSpecializedComponentType", Slang::StringUtil::makeStringWithFormat("0x%X", outDiagnosticsId));
+            }
+        }
+
+        m_fileStream.write(builder.begin(), builder.getLength());
+        m_fileStream.flush();
+    }
+
+    void CommonInterfaceWriter::link(ObjectID objectId, ObjectID outLinkedComponentTypeId, ObjectID outDiagnosticsId)
     {
+        Slang::StringBuilder builder;
+        int indent = 0;
+
+        Slang::String functionName = m_className;
+        functionName = functionName + "::link";
+
+        {
+            ScopeWritterForKey scopeWritter(&builder, &indent, functionName);
+            {
+                WRITE_PAIR(builder, indent, "this", Slang::StringUtil::makeStringWithFormat("0x%X", objectId));
+                WRITE_PAIR(builder, indent, "outLinkedComponentType", Slang::StringUtil::makeStringWithFormat("0x%X", outLinkedComponentTypeId));
+                WRITE_PAIR(builder, indent, "outDiagnostics", Slang::StringUtil::makeStringWithFormat("0x%X", outDiagnosticsId));
```

Comment:
Should the last one be using WRITE_PAIR_NO_COMMA?

Reply:
right, will change.